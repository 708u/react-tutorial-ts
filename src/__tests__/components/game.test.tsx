import React from 'react';
import renderer from 'react-test-renderer';
import { render, cleanup, screen, fireEvent } from '@testing-library/react';
import Game from 'components/game';

afterEach(cleanup);

describe('game component test', () => {
  it('should be same as a previous snapshot.', () => {
    const component = renderer.create(<Game />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should progress normal scenario', () => {
    render(<Game />);

    const playerXMsg = 'Next player: X';
    const playerOMsg = 'Next player: O';
    // first requirements.
    screen.getByText('Go to game start');
    screen.getByText(playerXMsg);
    expect(screen.queryByText(playerOMsg)).toBeNull();

    // progress first turn.
    fireEvent.click(screen.getByTestId('btn-0'));
    // reverse next player
    screen.getByText(playerOMsg);
    expect(screen.queryByText(playerXMsg)).toBeNull();
    // Go to this turn button was appeared.
    const expectedFirstHistoryStr = 'Go to move #1. [row: 1, col: 1]';
    screen.getByText(expectedFirstHistoryStr);

    // progress second turn.
    fireEvent.click(screen.getByTestId('btn-1'));
    // reverse next player
    screen.getByText(playerXMsg);
    expect(screen.queryByText(playerOMsg)).toBeNull();
    // Go to this turn button was appeared.
    const expectedSecondHistoryStr = 'Go to move #2. [row: 1, col: 2]';
    screen.getByText(expectedSecondHistoryStr);

    // time travel to first turn
    fireEvent.click(screen.getByText(expectedFirstHistoryStr));
    screen.getByText(playerOMsg);
    expect(screen.queryByText(playerXMsg)).toBeNull();
    // can go to current turn
    screen.getByText(expectedSecondHistoryStr);
    // button 1 value is gone.
    expect(screen.getByTestId('btn-1')).toBeEmptyDOMElement();
  });

  it('should win the X player', () => {
    render(<Game />);
    const clicks = ['btn-0', 'btn-7', 'btn-1', 'btn-8', 'btn-2'];
    // progress first turn.
    clicks.forEach((click) => fireEvent.click(screen.getByTestId(click)));

    // Winner comment was appeared.
    expect(screen.getByText('Winner: X')).toBeTruthy();
  });
});
