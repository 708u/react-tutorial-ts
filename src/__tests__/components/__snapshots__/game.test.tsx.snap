// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game component test should be same as a previous snapshot. 1`] = `
<div
  className="sc-gsTCUz cVJQdI"
>
  <div>
    <div
      className="sc-hKgILt fMCYZr"
    >
      <button
        className="sc-bdfBwQ iZlURE"
        data-testid="btn-0"
        name="1-1"
        onClick={[Function]}
      />
      <button
        className="sc-bdfBwQ iZlURE"
        data-testid="btn-1"
        name="1-2"
        onClick={[Function]}
      />
      <button
        className="sc-bdfBwQ iZlURE"
        data-testid="btn-2"
        name="1-3"
        onClick={[Function]}
      />
    </div>
    <div
      className="sc-hKgILt fMCYZr"
    >
      <button
        className="sc-bdfBwQ iZlURE"
        data-testid="btn-3"
        name="2-1"
        onClick={[Function]}
      />
      <button
        className="sc-bdfBwQ iZlURE"
        data-testid="btn-4"
        name="2-2"
        onClick={[Function]}
      />
      <button
        className="sc-bdfBwQ iZlURE"
        data-testid="btn-5"
        name="2-3"
        onClick={[Function]}
      />
    </div>
    <div
      className="sc-hKgILt fMCYZr"
    >
      <button
        className="sc-bdfBwQ iZlURE"
        data-testid="btn-6"
        name="3-1"
        onClick={[Function]}
      />
      <button
        className="sc-bdfBwQ iZlURE"
        data-testid="btn-7"
        name="3-2"
        onClick={[Function]}
      />
      <button
        className="sc-bdfBwQ iZlURE"
        data-testid="btn-8"
        name="3-3"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="sc-dlfnbm laRLSY"
  >
    <div
      className="sc-dlfnbm laRLSY"
    >
      Next player: X
    </div>
    <ol
      className="sc-eCssSg UPgMP"
    >
      <li>
        <button
          className="btn-bold"
          onClick={[Function]}
          type="button"
        >
          Go to game start
        </button>
      </li>
    </ol>
  </div>
</div>
`;
